[expo-type-information] update docs (#48322)
expo/expo: docs/pages/modules/type-generation-reference.mdx

- **view.tsx** for each view defined in the module
- **index.ts** file which reexports some functions

> If there are multiple modules provided, then a **Common.types.ts** file will be generated. It exports the common declarations (enums, structs), so they are defined exactly once.

Reviewer: ✏️ **style** `prose.md#voice-and-tone`

"will be generated" is future tense and passive. Name the tool as the actor and stay in present tense, like the rest of this section.

```suggestion
When you provide multiple modules, the tool also generates a **Common.types.ts** file. It exports the common declarations (enums and structs), so each one is defined exactly once.
```

<!-- docs-review -->

**Options:**
| Flag | Description | Default |
| :--- | :--- | :--- |
| `-a, --app-json <appJsonPath>` | A path to the [app config](/workflow/configuration/) file where the `expo.experiments.inlineModules.watchedDirectories` are defined. | |

> | `-w --watcher` | Starts a watcher that checks for changes in inline modules files. | |

Reviewer: 💡 **nit** `mechanics.md#check-these-vale-cannot-see-them`

Every other flag on this page now reads `-w, --watcher`, so this row is the last one still missing the comma.

```suggestion
| `-w, --watcher` | Starts a watcher that checks for changes in inline modules files. | |
```

<!-- docs-review -->

<Collapsible summary="Known issues about the current version of parsing the Swift file">

- Nested declarations will not be resolved fully.

>   `Sourcekitten` has a limit on resolving nested closures, so declarations nested deeper in your module (such as methods of a DSL `Class`) may not be fully resolved. Type inference doesn't work at this nesting depth, so only explicitly annotated types can be resolved there.

Reviewer: 💡 **nit** `terminology.md#external-product-casing`

The tool is spelled "SourceKitten", which is the spelling the options table on this page already uses.

```suggestion
  `SourceKitten` has a limit on resolving nested closures, so declarations nested deeper in your module (such as methods of a DSL `Class`) may not be fully resolved. Type inference doesn't work at this nesting depth, so only explicitly annotated types can be resolved there.
```

<!-- docs-review -->

- Unsupported Expo Module DSL declarations.

>   Most of the DSL declarations are parsed, but some details are not. For example, property setters (`Property(...).get { ... }.set { ... }`) are not modeled, so every generated property is `readonly`. Event payload types are not resolved, so generated event maps type the payload as a generic object.

Reviewer: ✏️ **style** `prose.md#voice-and-tone`

All three sentences are passive ("are parsed", "are not modeled", "are not resolved"). Make the tool the subject.

```suggestion
  The tool parses most DSL declarations, but not every detail. It doesn't model property setters (`Property(...).get { ... }.set { ... }`), so every generated property is `readonly`. It doesn't resolve event payload types either, so generated event maps type the payload as a generic object.
```

<!-- docs-review -->

- `SharedRef` types lose their `nativeRefType` tag.

>   A Swift `SharedRef<InnerType>` is generated as `SharedRef<any>` instead of `SharedRef<'nativeRefType'>`. The tool doesn't extract the `nativeRefType` string that a `SharedRef` subclass overrides (for example, `'image'` in `expo-image`), so the generated references are compatible with any `SharedRef` and are not checked against the tagged `SharedRef` types that other libraries declare.

Reviewer: ✏️ **style** `prose.md#sentence-construction`

The second sentence runs 39 words and carries three facts: what the tool skips, why, and what that means for type checking. Split it into three sentences.

```suggestion
  The tool generates a Swift `SharedRef<InnerType>` as `SharedRef<any>` instead of `SharedRef<'nativeRefType'>`. It doesn't extract the `nativeRefType` string that a `SharedRef` subclass overrides, such as `'image'` in `expo-image`. The generated references therefore work with any `SharedRef`, and are not checked against the tagged `SharedRef` types other libraries declare.
```

<!-- docs-review -->

- Using Unicode characters breaks `sourcekitten` offsets.

  By default the tool works around this by mapping all non-ASCII characters to ASCII strings before parsing. The `-s, --skip-unicode-character-mapping` option disables the workaround.

</Collapsible>
### Supported Expo modules DSL declarations

| Feature                      | Notes and limitations                                                                                                                                                             |
| :--------------------------- | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Expo DSL Declarations**    | Supports parsing for: `AsyncFunction`, `Class`, `Constant`, `Constructor`, `Events`, `Function`, `Name`, `Prop`, `Property`, `StaticAsyncFunction`, `StaticFunction`, and `View`. |

> | **Swift `struct` & `class`** | structs and classes conforming to `Record` protocol (technically classes/structs which declare `@Field`). Only properties marked with the `@Field` annotation are parsed.         |

Reviewer: ⚠️ **correctness** `prose.md#sentence-construction`

The parser treats a struct or class as a record when it declares at least one `@Field` **or** when it conforms to `Record` directly, so the two are alternatives rather than one being a technicality of the other. The cell also starts mid-sentence, unlike the two rows around it.

```suggestion
| **Swift `struct` & `class`** | Structs and classes that declare at least one `@Field`, or that conform to the `Record` protocol directly. Only properties marked with the `@Field` annotation are parsed.        |
```

<!-- docs-review -->

| **Swift `enum`**             | Basic cases are supported. Values associated with enum cases are not currently parsed.                                                                                            |

> ### Type conversions

Reviewer: ✏️ **style** `links-and-headings.md#titles-and-headings`

"Type conversions" and "Generated events" describe what the tool does, not what it fails to do, but as H3s the page outline files them under `## Swift parser limitations`. Promote both to H2 and move their subsections up one level.

<!-- docs-review -->

| `SharedRef<T>`                                  | `SharedRef<any>` (from `expo`)                                                   |

Types that cannot be mapped or resolved are generated as `unknown`. For type identifiers that are used but not declared in the parsed files, the tool additionally generates an `export type SomeType = unknown;` alias.

### Generated events

Events are generated differently depending on where they are declared:

- **Module events**: `Events(...)` in a module definition generates a `<ModuleName>Events` type, and the module class extends `NativeModule<<ModuleName>Events>`. Module events are sent with a dictionary body, so the payload is typed as `Record<string, any>`.
- **Class events**: `Events(...)` in a DSL `Class` generates a `<ClassName>Events` type, and the class extends `SharedObject<<ClassName>Events>`. Shared objects emit a single payload of any type, or none at all, so the listener is typed as `(payload?: any) => void`.
- **View events**: events declared in a `View` become optional callback props on the generated props interface, typed as `(event: any) => void`.

> The payload types of events are not resolved from the native code, which is why the payloads are typed permissively.

Reviewer: ✏️ **style** `review-ai-text.md#t5-restatement`

Each of the three bullets above already states how its payload is typed and why, and the DSL limitations section says it a third time. Cut this sentence.

<!-- docs-review -->
